### DISCUSSION TITLE
Can MOOSE compute integrals over a plane?

### DISCUSSION THREAD

**Discussion by alanchavez86** [idaholab/moose#17565] [Q&A General] [answered]
Hi all,

I'm still new to MOOSE and was curious as to how integrals are computed in moose. I understand that Finite element methods are used to numerically computed the solution to PDEs, but I'm interested in how moose goes about numerically computing integrals.

I've looked into the postprocessors section of the moose framework website for more information and found ElementAverageValue to be a possible solution to my question. 

Instead of using the Ferret approach, I'm instead seeing how I can simplify an integral for the demagnetizing field which influences decomposition behavior.

![image](https://user-images.githubusercontent.com/74686859/114242884-d67ad000-9948-11eb-8995-b6b8d0ec6025.png)

A simplified version for the integral part is given as an equation of phi:

![Screenshot 2021-04-09 153408](https://user-images.githubusercontent.com/74686859/114243002-10e46d00-9949-11eb-94a0-66f7d5ece997.png)

If rho, the magnetic charge density, is assumed to be 1, the integral be simplified in terms of x' and y' as:

![Screenshot 2021-04-09 153830](https://user-images.githubusercontent.com/74686859/114243444-d29b7d80-9949-11eb-93e4-a459f1e8dc44.png)

Would this be an appropriate approach to apply in MOOSE or would it be better off finding how to solve over a volume?

Thanks

Here's an additional document for further information (pg 2)
[Feng_2017_Peritectic PF Mag.pdf](https://github.com/idaholab/moose/files/6288636/Feng_2017_Peritectic.PF.Mag.pdf)

**Comment by GiudGiud** [comment]
Hello

moose will use the FE quadrature rules to compute the integral.
rho would be a variable right? And you want phi defined over the whole domain (so a variable) or just for a single position (more like a postprocessor) ?

For the latter, I would recommend looking at ElementIntegralVariablePostprocessor to do that. 
EDIT: Actually over a plane, maybe InterfaceIntegralVariablePostprocessor might be more appropriate. How is this plane defined in your mesh?
For the former, there may be a kernel already for this? @cticenhour will know more. 

Best,
Guillaume

**Reply by cticenhour** [reply]
@alanchavez86 I don't work on simulating ferromagnetic alloys, but I do recognize another form of these equations. They appear to be integral forms of a Poisson equation for the magnetic potential (also called the magnetostatic potential):

![image](https://user-images.githubusercontent.com/16690058/114419066-58990d80-9b70-11eb-86d2-a26b8517ea70.png)

and

![image](https://user-images.githubusercontent.com/16690058/114419686-e07f1780-9b70-11eb-8adc-4736e02189cc.png)

where M is the magnetization vector. I suppose you could solve for the magnetostatic potential as a variable using the `Diffusion` Kernel in the volume. But, that still requires knowledge of the right-hand-side magnetic charge density source term, as well as applicable boundary conditions on that potential. They solve for M in the paper using phase field, so I think you should look at the phase field module in conjunction with Eqn 3 in the paper to help get M. From there, you'll need M in an auxvariable in order to couple it into the poisson equation to calculate the gradient source term.

So that you don't go down too much of a rabbit hole with this, I'm also tagging @mangerij (as the lead of Ferret) who will have some specific knowledge in this area and may give you tips on alternative ways of solving this problem if you're wanting to forge your own path outside of the Ferret approach. I'll also tag @laagesen for general phase field.

**Reply by alanchavez86** [reply]
@GiudGiud Thanks for the reply and the insight, the plane is defined in my mesh as a 2D mesh:
[Mesh]
 type = GeneratedMesh
 dim = 2
 elem_type = QUAD4
 nx = 100
 ny = 100
 xmax = 200
 ymax = 200
[]
Thinking about this, I was confused whether or not solving the integral through x and y would be the correct approach and if it would be better solving it some other way.
I'm looking more into John and Casey's approach to implement the demagnetizing field into my input file.

**Comment by mangerij** [comment, marked as answer]
Yeah, I think they need **M** = **M**(c) where c is their concentration variable of the phase field problem. What confuses me is the paper I have been sent only gives an expression for M = M(c) as a _magnitude_ of the magnetization which makes it tricky to use in the Poisson equation since the righthand side is the divergence of the vector. At these temperatures, I highly doubt the vector direction matters...

But in principle if you calculate **M** correctly, Casey's suggestion is exactly what you need. Just add ∇⋅**M** as a Kernel. Here is what the residual entry would look like after integrating by parts

```
Real
PoissonRHS::computeQpResidual()
{
  RealVectorValue m(_mag_x[_qp], _mag_y[_qp], _mag_z[_qp]);
  return - m*_grad_test[_i][_qp];
}
```

**Reply by mangerij** [reply]
![poisson](https://user-images.githubusercontent.com/7597716/114534055-de8b8600-9c4e-11eb-8f02-8dcdb7972695.png)

this approach works with reasonable accuracy

**Reply by alanchavez86** [reply]
@mangerij Thank you for the response John. I was looking into creating a new kernel for ∇⋅M but had a quick question on how to add a C++ file to a MOOSE application. MOOSE mentions creating a C++ object, but I wasn't sure if I create that separately from moose or if you can create the C++ object from the moose application.

**Reply by mangerij** [reply]
You should add a .C file into src/kernels and an .h file into include/kernels. Both files should have the same name. Take a look how its done in any of the MOOSE examples. This is basically how you would set up an app.

**Reply by alanchavez86** [reply]
@mangerij Thank you for all the help! Sorry to bother you again, I was hoping you can take a look at a recent question I posted. I was hoping to get your thoughts on it. https://github.com/idaholab/moose/discussions/17899 Thank you.
